File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        self.graph = defaultdict(list)
        self.indegree = [0] * numCourses
        self.res = []

        def build_graph(preq):
            for v,u in preq:
                self.graph[u].append(v)
                self.indegree[v]+=1
    
        build_graph(prerequisites)

        queue = deque()
        for node in range(numCourses):
            if self.indegree[node]==0:
                queue.append(node)

        while queue:
            node = queue.popleft()
            self.res.append(node)

            for adj_node in self.graph[node]:
                self.indegree[adj_node]-=1
                if self.indegree[adj_node]==0:
                    queue.append(adj_node)
  
        return len(self.res)==numCourses

File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] indegree = new int[numCourses];
        List<List<Integer>> adj = new ArrayList<>(numCourses);
        
        for(int i=0; i<numCourses; i++){
            adj.add(new ArrayList<>());
        }
        
        for(int[] preq: prerequisites){
            adj.get(preq[1]).add(preq[0]);
            indegree[preq[0]]++;
        }
        
        Queue<Integer> queue = new LinkedList<>();
        
        for(int i=0; i<numCourses; i++){
            if(indegree[i]==0){
                queue.offer(i);
            }
        }
        
        int nodeVisited = 0;
        
        while (!queue.isEmpty()){
            int node = queue.poll();
            nodeVisited++;
            
            for(int nei:adj.get(node)){
                indegree[nei]--;
                if(indegree[nei]==0){
                    queue.offer(nei);
                }
            }
        }
        
        return nodeVisited == numCourses;
    }
}